Sources/MessagingPush/MessagingPush+AppDelegate.swift

```diff
@@ -71,20 +71,27 @@ extension MessagingPushImplementation {
                 // Customers with Universal Links in their app will need to add this function to their `AppDelegate` which will get called with deep link:
                 // func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void) -> Bool
 
-                var didHostAppHandleLink = false
+                // fallback to open link in device browser
+                let performDeepLinkFallback: () -> Void = {
+                    self.logger.debug("Host app didn't handle link. Opening the link through a system call.")
+                    UIApplication.shared.open(url: deepLinkurl)
+                }
+
+                logger.debug("Found a deep link inside of a push notification. \(deepLinkurl)")
+
```

Comment:
I am having a hard time reading and understanding the code and I believe it might get harder to even maintain it. What does breaking this into methods sound like? `Closure` is great, but they aren't easily maintainable and readable. 

And abstracting the deep link handling logic into a separate class or struct, making it easier to test and reuse.
 
```suggestion
if let deepLinkUrl = pushContent.deepLink {
    handleDeepLink(deepLinkUrl)
}

...

func handleDeepLink(_ deepLinkUrl: URL) {
    if deepLinkUtil.isLinkValidNSUserActivityLink(deepLinkUrl) {
        handleNSUserActivityLink(deepLinkUrl)
    } else {
        openLinkSystemWide(deepLinkUrl)
    }
}

func handleNSUserActivityLink(_ deepLinkUrl: URL) {
    let openLinkInHostAppActivity = NSUserActivity(activityType: NSUserActivityTypeBrowsingWeb)
    openLinkInHostAppActivity.webpageURL = deepLinkUrl

    let didHostAppHandleLink = UIApplication.shared.delegate?.application?(UIApplication.shared, continue: openLinkInHostAppActivity, restorationHandler: { _ in }) ?? false

    if !didHostAppHandleLink {
        openLinkSystemWide(deepLinkUrl)
    }
}

func openLinkSystemWide(_ deepLinkUrl: URL) {
    UIApplication.shared.open(url: deepLinkUrl)
}
```

Reply:
I like the idea of breaking the code apart into functions, but I avoided it because I do not like housing extra utility functions in this file. I like how this file focuses on being specific to `AppDelegate` functions and not utility methods. 

But, I do think it would look nice to leverage the `DeepLinkUtil` class more. 

This file could be modified to:

```swift
if let deepLinkUrl = pushContent.deepLink {
    deepLinkUtil.handleDeepLink(deepLinkUrl)
}
```

Then, the `DeepLinkUtil` could house the functions inside:
```swift
func handleDeepLink(_ deepLinkUrl: URL) {
   let ifHandled = handleUniversalLinkInHostApp(deepLinkUrl)

	if !ifHandled {
      openLinkSystemWide(deepLinkUrl)
    }
}
```

Reply:
That sounds good to me 👍